core/executor/src/wasm_executor.rs

```diff
@@ -679,6 +703,40 @@ impl WasmExecutor {
 		method: &str,
 		data: &[u8],
 	) -> Result<Vec<u8>> {
+		self.call_in_wasm_module_with_custom_signature(
+			ext,
+			module_instance,
+			method,
+			|alloc| {
+				let offset = alloc(data)?;
+				Ok(vec![I32(offset as i32), I32(data.len() as i32)])
+			},
+			|res, memory| {
+				if let Some(I64(r)) = res {
+					let offset = r as u32;
+					let length = (r >> 32) as u32 as usize;
```

Comment:
Why go through `u32`? Should be padded with zeros anyway, no?

Reply:
Could you explain more?

Reply:
I know that the code is copy pasted from some other place, so sorry for offtopic.

But my understanding is that we:
1. take the upper 32 bits of `r: i64`
2. Cast it to `u32`
3. Then (up)cast to `usize`
And the last step on 64-bit platforms will expand the value to `u64` by adding zeros.

My question was why we can't do:
```rust
(r >> 32) as usize
```
directly, since it should create a `i64` that contains upper bits shifted (so the original spots are now filled with 0s) and then cast to `usize`, which should pretty much be the same (unless I miss some casting caveats)

Reply:
Ohh yeah, it was to late yesterday and I was too tired. You are right and we can drop `as u32`.